Firestore/core/src/firebase/firestore/local/persistence.h

```diff
@@ -17,9 +17,13 @@
 #ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_LOCAL_PERSISTENCE_H_
 #define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_LOCAL_PERSISTENCE_H_
 
+#include <functional>
+#include <memory>
 #include <utility>
 
+#include "Firestore/core/src/firebase/firestore/local/reference_delegate.h"
```

Comment:
This seems like it can remain a forward declaration. Nothing from this type is actually used in this header.

Users of `persistence.h` that want to use something from the `reference_delegate()` should include `reference_delegate.h` for themselves.

Reply:
Done.

I did because compiler is complaining from `memory_persistence.h` for moving a forward declared class. Now in `memory_persistence.h` it is using include.

I have a question however: how to evaluate whether to include or forward declare in general? I feel like many headers could have used forward declaration by the reasoning you gave here: for example `remote_store.h` does not use anything from `local_store.h` so we could forward declare `LocalStore` there (but it is not).

I vaguely recall Costa said he preferred includes but I just could not recall his reason.

Reply:
The Google C++ Style Guide calls for [avoiding forward declarations](https://google.github.io/styleguide/cppguide.html#Forward_Declarations) and to use `#include` as much as possible. We initially followed this guidance.

The internal style guide justifies this on the grounds that with blaze and forge the compile time argument doesn't really matter so correctness and ease of use are the stronger argument.

I performed some experiments and found nontrivial compile-time improvements by using forward declarations where possible for just `util::Status` (though without changing structures). See #3761 for one change that resulted.

As a result, we've changed tacks. Compile time is a significant complaint from our users.